src/Riok.Mapperly/Descriptors/MappingBuilders/DictionaryMappingBuilder.cs

```diff
@@ -66,6 +75,10 @@ public static class DictionaryMappingBuilder
         if (BuildKeyValueMapping(ctx) is not var (keyMapping, valueMapping))
             return null;
 
+        // if target is an immutable dictionary then don't create a foreach loop
+        if (ctx.Target.OriginalDefinition.ImplementsGeneric(ctx.Types.IImmutableDictionaryT, out _))
+            return null;
```

Comment:
I think a diagnostic here would be super helpful, we recently implemented  a similar thing for enumerables: 
https://github.com/riok/mapperly/pull/345/files#diff-d0e8d4a6f1b4bc7cabf6bf20484b1ff8e57a8fde79753fa4bf2d563df38965dbR94-R97